extra/triggers.rst

```diff
@@ -0,0 +1,204 @@
+  SELECT create_distributed_table('big_vals',    'key');
+
+  -- This trigger function takes the destination placement as an argument
+
+  SELECT run_command_on_workers($cmd$
+    CREATE OR REPLACE FUNCTION embiggen() RETURNS TRIGGER AS $$
+      BEGIN
+        IF (TG_OP = 'INSERT') THEN
+          EXECUTE format(
+            'INSERT INTO %s (key, val) SELECT ($1).key, ($1).val*2;',
+            TG_ARGV[0]
+          ) USING NEW;
+        END IF;
+        RETURN NULL;
+      END;
+    $$ LANGUAGE plpgsql;
+  $cmd$);
+
+  -- Next we relate the co-located tables by the trigger function
+  -- on each co-located placement
+
+  SELECT run_command_on_colocated_placements(
+    'little_vals',
+    'big_vals',
+    $cmd$
+      CREATE TRIGGER after_insert AFTER INSERT ON %s
+        FOR EACH ROW EXECUTE PROCEDURE embiggen(%s)
+    $cmd$
+  );
+
+Then to test it:
+
+.. code-block:: postgresql
+
+  INSERT INTO little_vals VALUES (1, 42), (2, 101);
+  TABLE big_vals;
+
+::
+
+  ┌─────┬─────┐
+  │ key │ val │
+  ├─────┼─────┤
+  │   1 │  84 │
+  │   2 │ 202 │
+  └─────┴─────┘
+
+**Trigger between reference tables.**
+
+.. note::
+
+  **This workaround is only safe in limited situations.** When using such a trigger to insert into a reference table, make sure that no concurrent updates happen on the destination table. The order in which concurrent update/delete/insert commands are applied to replicas is not guaranteed, and replicas of the reference table can get out of sync with one another.
+
+Reference tables are simpler than distributed tables in that they have exactly one shard which is replicated across all workers. To relate reference tables with a trigger, we can create a trigger for the shard on all workers.
+
+Suppose we want to record the author of every change in ``insert_target`` to ``audit_table``, both of which are reference tables. As long as nothing but our trigger updates the ``audit_table`` then this will be safe.
+
+.. code-block:: postgresql
+
+  -- create the reference tables
+
+  CREATE TABLE insert_target (
+    value text
+  );
+  CREATE TABLE audit_table(
+    author name NOT NULL,
+    value text
+  );
+  SELECT create_reference_table('insert_target');
+  SELECT create_reference_table('audit_table');
+
+  -- now find the shard id for each table
+
+  SELECT logicalrelid, shardid
+  FROM pg_dist_shard
+  WHERE logicalrelid IN
+    ('insert_target'::regclass, 'audit_table'::regclass);
+
+Include the shard ids (written below as "xxxxxx" and "yyyyyy") in custom queries:
+
+.. code-block:: postgresql
+
+  SELECT run_command_on_workers($cmd$
+    CREATE OR REPLACE FUNCTION process_audit() RETURNS TRIGGER AS $$
+      BEGIN
+        INSERT INTO audit_table_xxxxxx(author,value)
+          VALUES (current_user,NEW.value);
+        RETURN NEW;
+      END;
+    $$ LANGUAGE plpgsql;
+  $cmd$);
+
+  SELECT run_command_on_workers($cmd$
+    CREATE TRIGGER emp_audit
+    AFTER INSERT OR UPDATE ON insert_target_yyyyyy
```

Comment:
I'm still not sure why we need shard-id's for both reference tables? We do need it for the destination table, but the source table shouldn't be needed. Can we do run_command_on_shards (or placements) instead?

Reply:
The tables `insert_target` and `audit_table` don't exist on the workers. Referencing either of those tables within `run_command_on_workers()` will cause a "relation does not exist" error.

For instance:

```sql
-- produces a value
select count(*) from insert_target;

-- no good
select run_command_on_workers('select count(*) from insert_target');
┌────────────────────────────────────────────────────────────────────────┐
│                         run_command_on_workers                         │
├────────────────────────────────────────────────────────────────────────┤
│ (localhost,5433,f,"ERROR:  relation ""insert_target"" does not exist") │
│ (localhost,5434,f,"ERROR:  relation ""insert_target"" does not exist") │
└────────────────────────────────────────────────────────────────────────┘

-- but this works
select run_command_on_workers('select count(*) from insert_target_102425');
```

Reply:
Yes, but run_command_on_shards (or placements) should resolve the shard_id for the table. Why can't we use the other API's? Do we have to use the generic run_command_on_workers?

Reply:
I see what you're saying. I'll try to simplify it with one of the other UDFs.